common4j/src/main/com/microsoft/identity/common/java/net/UrlConnectionHttpClient.java

```diff
@@ -0,0 +1,361 @@
+        urlConnection.setRequestMethod(request.getRequestMethod());
+        urlConnection.setUseCaches(true);
+        setRequestBody(urlConnection, request.getRequestContent(), request.getRequestHeaders().get(HttpConstants.HeaderField.CONTENT_TYPE));
+
+        InputStream responseStream = null;
+
+        HttpResponse response = null;
+        try {
+            try {
+                responseStream = urlConnection.getInputStream();
+            } catch (final SocketTimeoutException socketTimeoutException) {
+                // SocketTimeoutExcetion is thrown when connection timeout happens. For connection
+                // timeout, we want to retry once. Throw the exception to the upper layer, and the
+                // upper layer will handle the retry.
+                throw socketTimeoutException;
+            } catch (final IOException ioException) {
+                // 404, for example, will generate an exception.  We should catch it.
+                responseStream = urlConnection.getErrorStream();
+            }
+
+            final int statusCode = urlConnection.getResponseCode();
+            final Date date = new Date(urlConnection.getDate());
+
+            final String responseBody = responseStream == null
+                    ? ""
+                    : convertStreamToString(responseStream);
+
+            response = new HttpResponse(
+                    date,
+                    statusCode,
+                    responseBody,
+                    urlConnection.getHeaderFields()
+            );
+        } finally {
+            completionCallback.accept(response);
+            safeCloseStream(responseStream);
+        }
+
+        return response;
+    }
+
+    private HttpURLConnection setupConnection(HttpRequest request) throws IOException {
+        final HttpURLConnection urlConnection = HttpUrlConnectionFactory.createHttpURLConnection(request.getRequestUrl());
+
+        // Apply request headers and update the headers with default attributes first
+        final Set<Map.Entry<String, String>> headerEntries = request.getRequestHeaders().entrySet();
+
+        for (final Map.Entry<String, String> entry : headerEntries) {
+            urlConnection.setRequestProperty(entry.getKey(), entry.getValue());
+        }
+
+        urlConnection.setConnectTimeout(getConnectTimeoutMs());
+        urlConnection.setReadTimeout(getReadTimeoutMs());
+        urlConnection.setInstanceFollowRedirects(true);
+        urlConnection.setUseCaches(false);
+        urlConnection.setDoInput(true);
+
+        return urlConnection;
+    }
+
+    private Integer getReadTimeoutMs() {
+        return readTimeoutMsSupplier == null ? readTimeoutMs : readTimeoutMsSupplier.get();
+    }
+
+    private Integer getConnectTimeoutMs() {
+        return connectTimeoutMsSupplier == null ? connectTimeoutMs : connectTimeoutMsSupplier.get();
+    }
+
+    private static void setRequestBody(@NonNull final HttpURLConnection connection,
```

Comment:
>setRequestBody [](http://example.com/codeflow?start=24&length=14)

I guess this method also sends the request, right?
in this case we should rename the method name..